tests/cupy_tests/creation_tests/test_from_data.py

```diff
@@ -475,24 +439,80 @@ def test_asfortranarray_cuda_array_zero_dim_dtype(
         return xp.asfortranarray(a, dtype=dtype_b)
 
 
-class DummyObjectWithCudaArrayInterface(object):
+max_cuda_array_interface_version = 2
+
 
-    def __init__(self, a, mask=None):
+@testing.gpu
+@testing.parameterize(*testing.product({
+    'ver': tuple(range(max_cuda_array_interface_version+1)),
+}))
+class TestCudaArrayInterface(unittest.TestCase):
+    @testing.for_all_dtypes()
+    def test_base(self, dtype):
+        a = testing.shaped_arange((2, 3, 4), cupy, dtype)
+        b = cupy.asarray(DummyObjectWithCudaArrayInterface(a, self.ver))
+        testing.assert_array_equal(a, b)
+
+    @testing.for_all_dtypes()
+    def test_not_copied(self, dtype):
+        a = testing.shaped_arange((2, 3, 4), cupy, dtype)
+        b = cupy.asarray(DummyObjectWithCudaArrayInterface(a, self.ver))
+        a.fill(0)
+        testing.assert_array_equal(a, b)
+
+    @testing.for_all_dtypes()
+    def test_order(self, dtype):
+        a = testing.shaped_arange((2, 3, 4), cupy, dtype)
+        b = cupy.asarray(DummyObjectWithCudaArrayInterface(a, self.ver),
+                         order='F')
+        assert b.flags.f_contiguous
+        testing.assert_array_equal(a, b)
+
+    @testing.for_all_dtypes()
+    def test_with_strides(self, dtype):
+        a = testing.shaped_arange((2, 3, 4), cupy, dtype).T
+        b = cupy.asarray(DummyObjectWithCudaArrayInterface(a, self.ver))
+        assert a.strides == b.strides
+        assert a.nbytes == b.data.mem.size
+
+
+@testing.gpu
+@testing.parameterize(*testing.product({
+    'ver': tuple(range(1, max_cuda_array_interface_version+1)),
+}))
+class TestCudaArrayInterfaceMaskedArray(unittest.TestCase):
+    # TODO(leofang): update this test when masked array is supported
+    @testing.for_all_dtypes()
+    def test_masked_array(self, dtype):
+        a = testing.shaped_arange((2, 3, 4), cupy, dtype)
+        mask = testing.shaped_arange((2, 3, 4), cupy, dtype)
+        a = DummyObjectWithCudaArrayInterface(a, self.ver, mask)
+        with pytest.raises(ValueError) as ex:
+            b = cupy.asarray(a)  # noqa
+        assert 'does not support' in str(ex.value)
+
+
+class DummyObjectWithCudaArrayInterface(object):
+    def __init__(self, a, ver, mask=None):
+        assert ver in tuple(range(max_cuda_array_interface_version+1))
         self.a = a
+        self.ver = ver
         self.mask = mask
 
     @property
     def __cuda_array_interface__(self):
         desc = {
             'shape': self.a.shape,
-            'strides': self.a.strides,
             'typestr': self.a.dtype.str,
             'descr': self.a.dtype.descr,
             'data': (self.a.data.ptr, False),
-            'version': 2,
+            'version': self.ver,
         }
-        if self.mask is not None:
-            desc['mask'] = self.mask
+        if self.ver == 2 or self.a._f_contiguous:
```

Comment:
Instead of determining by the version, how about passing a boolean flag `include_strides` or something?
This test configuration is missing:
* version is either 0 or 1, and
* array is contiguous, and
* strides is included.

Reply:
Check it out 06012a2. I think it provides a better coverage with more parametrizations.